lib/trln_argon/mappings.rb

```diff
@@ -0,0 +1,150 @@
+    end
+
+    def refresh
+      if File.directory?(File.join(@repo_dir, '.git'))
+        @git ||= Git.init(@repo_dir)
+        @git.pull('origin', @branch)
+      else
+        clone
+      end
+    end
+  end
+
+  class Lookups
+    attr_reader :directory
+
+    KEYS = {
+      library: 'library location',
+      location: 'shelving location'
+    }.freeze
+
+    PATH_COMPONENTS = %i[library location].freeze
+
+    FILENAMES = {
+      location_holdings: 'location_item_holdings.json',
+      location_facet: 'location_facet.json'
+    }.freeze
+
+    def initialize(base = '.')
+      @directory = base
+      reload!
+    end
+
+    # looks up a display value given a path of the form
+    # "[inst_code].[lookup_type].[code]", e.g.
+    # `unc.location_facet.uncgrar' looks up the code to be used when displaying
+    # the location facet
+    def lookup(path)
+      parts = path.split('.')
+      ctx = @mappings
+      parts.each do |k|
+        ctx = ctx[k]
+        break if ctx.nil? || ctx.empty?
+      end
+      ctx.nil? || ctx.empty? ? path : ctx
+    end
+
+    def mappings
+      @mappings ||= load
+    end
+
+    def reload!
+      @mappings = load
+    end
+
+    def load
+      mappings = {}
+      Dir.foreach(@directory) do |dir_entry|
+        path = File.expand_path(File.join(@directory, dir_entry))
+        next unless File.directory?(path) && dir_entry =~ /^[a-z]/
+        inst_mappings = mappings[File.basename(path)] = {}
+        lhf = File.join(path, FILENAMES[:location_holdings])
+        parse_holdings!(lhf, inst_mappings)
+        lff = File.join(path, FILENAMES[:location_facet])
+        facets = Hash.new do |hash, key|
+          hash[key] = inst_mappings['library'][key]
+        end
+        facets.update(read_json(lff))
+        inst_mappings['facet'] = facets
+      end
+      mappings
+    end
+
+    private
+
+    def parse_holdings!(filename, inst_mappings)
+      lookups = read_json(filename)
+
+      library_mappings = (inst_mappings['library'] ||= {})
+      libraries = lookups.fetch(KEYS[:library], {})
+      library_mappings.update(libraries)
+
+      locations = lookups.fetch(KEYS[:location], {})
+      location_mappings = (inst_mappings['location'] ||= {})
+      location_mappings.update(locations)
+    end
+
+    def read_json(filename)
+      File.exist?(filename) ? File.open(filename) { |f| JSON.parse(f.read) } : {}
+    end
+  end
+
+  # Mappings for library/location names, statuses, etc.
+  class LookupManager
+    include Singleton
+
+    CACHE_KEY = 'TrlrArgon::LookupManager::Lookups'.freeze
+
+    class << self
+      attr_writer :fetcher
+
+      def fetcher
+        @fetcher ||= TrlnArgon::MappingsGitFetcher.new
+      end
+    end
+
+    def reload
+      self.class.fetcher.refresh
+      Rails.cache.delete(CACHE_KEY)
+    end
+
+    def map(path)
+      lookups.lookup(path)
+    end
+
+    def lookups
+      Rails.cache.fetch(CACHE_KEY, expires_in: 24.hours) do
```

Comment:
@adjam this won't work when caching is on, I think because of the Hash.new block.

https://stackoverflow.com/questions/6391855/rails-cache-error-in-rails-3-1-typeerror-cant-dump-hash-with-default-proc#answer-6392704